corehq/apps/userreports/reports/filters/choice_providers.py

```diff
@@ -25,34 +23,77 @@ def offset(self):
         return self.page * self.limit
 
 
-def get_choices_from_data_source_column(query_context):
-    adapter = IndicatorSqlAdapter(query_context.report.config)
-    table = adapter.get_table()
-    sql_column = table.c[query_context.report_filter.field]
-    query = adapter.session_helper.Session.query(sql_column)
-    if query_context.query:
-        query = query.filter(sql_column.contains(query_context.query))
+class ChoiceProvider(object):
+    __metaclass__ = ABCMeta
+
+    def __init__(self, report, filter_slug):
+        self.report = report
+        self.filter_slug = filter_slug
+
+    @property
+    def report_filter(self):
+        return self.report.get_ui_filter(self.filter_slug)
+
+    @property
+    def domain(self):
+        return self.report.domain
+
+    @abstractmethod
+    def query(self, query_context):
+        pass
+
+    @abstractmethod
+    def get_choices_for_values(self, values):
+        pass
+
+
+class DataSourceColumnChoiceProvider(ChoiceProvider):
 
-    try:
+    def query(self, query_context):
+        return self.get_choices_for_values(self.get_values_for_query(query_context))
+
+    @property
+    def _adapter(self):
+        return IndicatorSqlAdapter(self.report.config)
+
+    @property
+    def _sql_column(self):
+        return self._adapter.get_table().c[self.report_filter.field]
+
+    def get_values_for_query(self, query_context):
+        query = self._adapter.session_helper.Session.query(self._sql_column)
+        if query_context.query:
+            query = query.filter(self._sql_column.contains(query_context.query))
+
+        query = query.distinct().order_by(self._sql_column).limit(query_context.limit).offset(query_context.offset)
+        try:
+            return [v[0] for v in query]
+        except ProgrammingError:
+            return []
+
+    def get_choices_for_values(self, values):
+        return [Choice(value, value) for value in values]
+
+
+class LocationChoiceProvider(ChoiceProvider):
```

Comment:
Dang, I was sure that I wouldn't release that without tests but i guess not https://github.com/dimagi/commcare-hq/pull/8947/. The dangers of time-sensitive client work (and a lot of bootstrap code needed for tests).

The easiest thing to do is to setup a location hierarchy in a project locally and then add a location-based data source (see `LOCATIONS_IN_UCR` toggle) and a report filter like the following:

```
[
  {
    "compare_as_string": false, 
    "datatype": "string", 
    "choice_provider": {
      "type": "location"
    }, 
    "display": "Location", 
    "field": "doc_id", 
    "type": "dynamic_choice_list", 
    "slug": "doc_id"
  }
]
```

if you were brave enough to write tests you could look at other commtrack tests that bootstrap locations.